storage/innobase/lock/lock0lock.cc

```diff
@@ -4501,7 +4503,9 @@ static void lock_rec_unlock_unmodified(hash_cell_t &cell, lock_t *lock,
   {
     if (UNIV_UNLIKELY(!page_is_leaf(block->page.frame)))
     {
-      ut_ad("corrupted lock system" == 0);
+      /* There can be locks with clean bitmap on non-leaf pages in the case of
+      page split. See lock_move_rec_list_end() call stack for details. */
+      ut_ad(lock_rec_find_set_bit(lock) == ULINT_UNDEFINED);
```

Comment:
Yes, this test is innodb.lock_update_split_rc. 

The stack trace is the following:
``` text
#0  0x00005ebaaabe8da7 in lock_rec_reset_nth_bit (lock=0x7244198d7d00, i=4) at ./storage/innobase/include/lock0priv.h:489
#1  0x00005ebaaabd1041 in lock_move_rec_list_end (new_block=0x7244190e6700, block=0x7244190e6490, rec=0x7244195a8063 "infimum")
    at ./storage/innobase/lock/lock0lock.cc:2879                                
#2  0x00005ebaaac82a02 in page_copy_rec_list_end (new_block=0x7244190e6700, block=0x7244190e6490, rec=0x7244195a8063 "infimum", index=0x7243f000de68,
    mtr=0x7244111a8db0, err=0x7244111a893c) at ./storage/innobase/page/page0page.cc:663
#3  0x00005ebaaae17284 in btr_root_raise_and_insert (flags=0, cursor=0x7244111a8a80, offsets=0x7244111a8a00, heap=0x7244111a89f8, tuple=0x7243fc036df8,
    n_ext=0, mtr=0x7244111a8db0, err=0x7244111a893c) at ./storage/innobase/btr/btr0btr.cc:1987
#4  0x00005ebaaae44049 in btr_cur_pessimistic_insert (flags=0, cursor=0x7244111a8a80, offsets=0x7244111a8a00, heap=0x7244111a89f8, entry=0x7243fc036df8,
    rec=0x7244111a8a08, big_rec=0x7244111a89f0, n_ext=0, thr=0x7243fc036298, mtr=0x7244111a8db0)
    at ./storage/innobase/btr/btr0cur.cc:2754                                   
#5  0x00005ebaaacfbfff in row_ins_clust_index_entry_low (flags=0, mode=BTR_MODIFY_TREE, index=0x7243f000de68, n_uniq=1, entry=0x7243fc036df8, n_ext=0,
    thr=0x7243fc036298) at ./storage/innobase/row/row0ins.cc:2855               
#6  0x00005ebaaacfd6c1 in row_ins_clust_index_entry (index=0x7243f000de68, entry=0x7243fc036df8, thr=0x7243fc036298, n_ext=0)
    at ./storage/innobase/row/row0ins.cc:3256
```

It looks like it was done intentionally. Here is the comment of lock_update_root_raise() function:
```text
/** Update locks when the root page is copied to another in                     
btr_root_raise_and_insert(). Note that we leave lock structs on the             
root page, even though they do not make sense on other than leaf                
pages: the reason is that in a pessimistic update the infimum record            
of the root page will act as a dummy carrier of the locks of the record         
to be updated. */                                                               
void lock_update_root_raise(const buf_block_t &block, const page_id_t root)
```

There is also lock_update_copy_and_discard() function, which discards lock objects from the certain page, and it's used for merging pages, but it's not used in the above call stack.

It does not cause lock objects leak because the lock objects will be released on transaction commit or rollback.